File: frontend/src/types/sale.ts
export interface SaleItemRequest {
  itemName: string;
  quantity: number;
  unitPrice: number;
}

export interface SaleRequest {
  saleType: string;
  customerId?: number;
  totalAmount: number;
  paidAmount: number;
  saleDate?: string;
  notes?: string;
  items: SaleItemRequest[];
}

export interface SaleItemResponse {
  id: number;
  itemName: string;
  quantity: number;
  unitPrice: number;
  totalPrice: number;
}

export interface SaleResponse {
  id: number;
  customerId?: number;
  customerName?: string;
  saleType: string;
  totalAmount: number;
  paidAmount: number;
  balanceAmount: number;
  saleDate: string;
  notes?: string;
  items: SaleItemResponse[];
  createdAt: string;
  updatedAt: string;
}

File: backend/src/main/java/com/vyapaarbuddy/entity/Sale.java
package com.vyapaarbuddy.entity;

import com.vyapaarbuddy.enums.SaleType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

/**
 * Sale entity representing sales transactions.
 * TODO: Add invoice number generation
 * TODO: Add payment method tracking
 * TODO: Add delivery status
 */
@Entity
@Table(name = "sales")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Sale {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    @Column(nullable = false)
    private LocalDate saleDate;

    @Enumerated(EnumType.STRING)
    private SaleType type;

    @Column(precision = 15, scale = 2, nullable = false)
    private BigDecimal totalAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal discountAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal taxAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal paidAmount;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Builder.Default
    @OneToMany(mappedBy = "sale", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<SaleItem> items = new HashSet<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_id", nullable = false)
    private Business business;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
